File: src/utils/uds_service.cpp
#include "utils/uds_service.hpp"

UDSService::UDSService(const std::string& local_path, const std::string& receiver_path)
    : local_path_(local_path), receiver_path_(receiver_path) {
    sock_ = socket(AF_UNIX, SOCK_DGRAM, 0);
    if (sock_ < 0)
        throw std::runtime_error("Socket creation failed.");

    memset(&local_addr_, 0, sizeof(local_addr_));
    local_addr_.sun_family = AF_UNIX;
    std::strncpy(local_addr_.sun_path, local_path_.c_str(), sizeof(local_addr_.sun_path) - 1);
    ::unlink(local_path_.c_str());  // Remove stale socket file if it exists
    if (bind(sock_, reinterpret_cast<struct sockaddr*>(&local_addr_), sizeof(local_addr_)) < 0) {
        close(sock_);
        throw std::runtime_error("Bind failed.");
    }

    memset(&receiver_addr_, 0, sizeof(receiver_addr_));
    receiver_addr_.sun_family = AF_UNIX;
    std::strncpy(receiver_addr_.sun_path, receiver_path_.c_str(), sizeof(receiver_addr_.sun_path) - 1);
}

void UDSService::sendMessage(const std::string& message) {
    ssize_t sent = sendto(sock_, message.c_str(), message.length(), 0,
                          reinterpret_cast<struct sockaddr*>(&receiver_addr_), sizeof(receiver_addr_));
    if (sent < 0)
        throw std::runtime_error("Send failed.");
}

std::string UDSService::receiveMessage() {
    char buffer[512];
    memset(buffer, 0, sizeof(buffer));
    ssize_t recv_len = recvfrom(sock_, buffer, sizeof(buffer) - 1, 0, nullptr, nullptr);
    if (recv_len < 0)
        throw std::runtime_error("Receive failed.");

    buffer[recv_len] = '\0';
    return buffer;
}

File: src/query/variable_order/uds_service.py
import atexit
import os
import socket

class UDSService:
    def __init__(self, local_path: str, receiver_path: str):
        self.receiver_path = receiver_path
        self.local_path = local_path  # 本地监听的 Unix 域套接字路径
        self.buffer_size = 1024

        if os.path.exists(self.local_path):
            os.remove(self.local_path)

        # 创建 Unix Domain Datagram 套接字
        self.udp_socket = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)

        # 绑定本地地址到套接字（用于接收消息）
        self.udp_socket.bind(self.local_path)
        atexit.register(self.close)
        print(f"UDS server is listening on {self.local_path}")

    def receive_message(self):
        data, _ = self.udp_socket.recvfrom(self.buffer_size)
        return data.decode()

    def send_message(self, message):
        self.udp_socket.sendto(message.encode(), self.receiver_path)

    def close(self):
        try:
            self.udp_socket.close()
        finally:
            if os.path.exists(self.local_path):
                os.remove(self.local_path)
